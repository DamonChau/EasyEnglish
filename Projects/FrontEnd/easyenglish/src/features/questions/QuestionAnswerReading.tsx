/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { useState, useEffect } from "react";
import { QuestionsResponse, useGetQuestionsWithQDQuery } from "./questionsApi";
import { useAddUserAnswerMutation } from "../users/userAnswersApi";
import { useAddExamResultMutation } from "../examResult/examResultApi";
import { useTypedSelector } from "../../services";
import isUUID from "validator/lib/isUUID";
import {
  Questions,
  UserAnswersDisplay,
  QuestionDetails,
  UserAnswers,
  ExamResults,
} from "../../interfaces/interfaces";
import {
  selectLoggedUser,
  selectIsAuthenticated,
} from "../../services/slices/authSlice";
import Snackbar from "@mui/material/Snackbar";
import Drawer from "@mui/material/Drawer";
import { useStateWithCallbackLazy } from "use-state-with-callback";
import { v4 as uuidv4 } from "uuid";
import { ListAnswerResult } from "./QuestionAnswerReadingResult";
import QuestionAnswerReadingDetails from "./QuestionAnswerReadingDetails";
import { Alert } from "../common/Modals";

const QuestionAnswerReading = ({ testId, getExamResult }: any) => {
  const [isView, setView] = React.useState(false);
  const { data, isFetching, isLoading, isSuccess, isError, error } =
    useGetQuestionsWithQDQuery(testId, { skip: !isView });
  const [
    addUserAnswer,
    { isLoading: isAddUALoading, isError: isAddUAError, error: errorUAAdd },
  ] = useAddUserAnswerMutation();
  const [
    addExamResult,
    { isLoading: isAddERLoading, isError: isAddERError, error: errorERAdd },
  ] = useAddExamResultMutation();

  const [userAnswers, setUserAnswers] = useStateWithCallbackLazy<
    UserAnswersDisplay[]
  >([]);
  const [open, setOpen] = React.useState(false);
  const [openDrawer, setOpenDrawer] = React.useState(false);
  const [isSave, setSave] = React.useState(false);
  const loggedUser = useTypedSelector(selectLoggedUser);
  const isAuthenticated = useTypedSelector(selectIsAuthenticated);
  const initialValueTestResult = {
    id: uuidv4(),
    examTestId: testId,
    score: 0,
    noQuestion: 0,
    createdBy: isAuthenticated ? loggedUser!.id : uuidv4(),
  };
  const [testResult, setTestResult] = useState(initialValueTestResult);
  const [erroMsg, setErrorMsg] = useState("");

  //update exam test result id
  useEffect(() => {
    if (testId && isUUID(testId)) {
      setView(true);
      setTestResult((prevState) => ({ ...prevState, examTestId: testId }));
    }
  }, [testId]);

  //update exam test result no question
  useEffect(() => {
    if (data) {
      const questions: Questions[] = data;
      questions.forEach((q) => {
        setTestResult((prevState) => ({
          ...prevState,
          noQuestion: prevState.noQuestion + q.questionDetails.length,
        }));
      });
    }
  }, [data]);

  useEffect(() => {
    if (error) {
      if ("status" in error) {
        // you can access all properties of `FetchBaseQueryError` here
        const msg = "error" in error ? error.error : JSON.stringify(error.data);
        setErrorMsg(msg);
      } else {
        // you can access all properties of `SerializedError` here
        setErrorMsg(error.message as string);
      }
    }
  }, [isError]);

  //#region helpers
  function replaceWithBr(q: Questions) {
    return q?.content.replace(/\n/g, "<br />");
  }

  const findAnswer = (
    questions: QuestionsResponse,
    questionDetailId: string
  ): QuestionDetails => {
    let questionDetail: QuestionDetails = {
      id: "",
      order: -1,
      answer: "",
      content: "",
      createdDate: "",
      createdBy: "",
      questionId: "",
    };
    for (const q of questions) {
      for (const qd of q.questionDetails) {
        if (qd.id === questionDetailId) {
          questionDetail = qd;
          break;
        }
      }
    }
    return questionDetail;
  };

  const compareFB = (a: UserAnswersDisplay, b: UserAnswersDisplay) => {
    if (a.order < b.order) {
      return -1;
    }
    if (a.order > b.order) {
      return 1;
    }
    return 0;
  };
  //#endregion

  //#region events
  const handleFieldChange = React.useCallback(
    (userAnswers: UserAnswersDisplay) => {
      setUserAnswers(
        (prevState) => {
          // Check if the object already exists in the array
          const existingObjectIndex = prevState.findIndex(
            (obj) =>
              obj.questionDetailId === userAnswers.questionDetailId &&
              obj.userId === userAnswers.userId
          );

          // If it exists, update the object at the existing index
          if (existingObjectIndex !== -1) {
            const updatedArray = [...prevState];
            updatedArray[existingObjectIndex] = userAnswers;
            return updatedArray;
          }

          // If it doesn't exist, add the new object to the end of the array
          return [...prevState, userAnswers];
        },
        () => {}
      );
    },
    []
  );

  const handleClose = (
    event?: React.SyntheticEvent | Event,
    reason?: string
  ) => {
    if (reason === "clickaway") {
      return;
    }

    setOpen(false);
  };

  const toggleDrawer = () => {
    setOpenDrawer(!openDrawer);
  };

  const restartNewTest = () => {
    setUserAnswers([], () => {});
    setOpen(false);
    setOpenDrawer(false);
    setSave(false);
    setTestResult(initialValueTestResult);
    setView(true);
  };
  //#endregion

  //#region save and check for result
  const onSaveExamResult = async (
    result: Partial<ExamResults>,
    score: number
  ) => {
    try {
      if (result) {
        result.score = score;
        const r = await addExamResult(result).unwrap();
        setTestResult(r);
        return r.id;
      }
    } catch (error) {
      console.log(error);
    }
  };

  const onSaveUserAnswer = async (userAnswers: UserAnswersDisplay[]) => {
    try {
      if (userAnswers) {
        userAnswers.forEach(async (answer: Partial<UserAnswers>) => {
          const r = await addUserAnswer(answer).unwrap();
          setUserAnswers(
            (prevState) => {
              // Check if the object already exists in the array
              const existingObjectIndex = prevState.findIndex(
                (obj) =>
                  obj.questionDetailId === r.questionDetailId &&
                  obj.userId === r.userId &&
                  obj.examResultId == r.examResultId
              );

              // If it exists, update the object at the existing index
              if (existingObjectIndex !== -1) {
                const updatedArray = [...prevState];
                updatedArray[existingObjectIndex].id = r.id;

                return updatedArray;
              } else {
                const updatedArray = [...prevState];
                return updatedArray;
              }
            },
            () => {}
          );
        });
      }
    } catch (error) {
      console.log(error);
    }
  };

  const onSave = async () => {
    try {
      let score = 0;
      //update score and result
      let updatedUserAnswers = userAnswers.map((answer) => {
        const qd = findAnswer(
          data as QuestionsResponse,
          answer.questionDetailId
        );
        answer.answer == qd.answer ? (answer.result = 1) : (answer.result = 0);
        answer.answerOrg = qd.answer;
        answer.order = qd.order;
        answer.description = qd.order.toString();
        score += answer.result;
        return answer;
      });
      //save ExamResult
      const examResultId = await onSaveExamResult(testResult, score);

      //save User Answer
      updatedUserAnswers = userAnswers.map((answer) => {
        // Create a new object with the updated id value
        const updatedAnswer = { ...answer, examResultId: examResultId! };
        return updatedAnswer;
      });
      updatedUserAnswers.sort(compareFB);

      setUserAnswers(updatedUserAnswers, () => {
        onSaveUserAnswer(updatedUserAnswers);
      });

      setOpen(true);
      toggleDrawer();
      setSave(true);
      setView(false);
    } catch (error) {
      console.log(error);
      setOpen(false);
      setSave(false);
    }
  };

  const onCreateNewNote = () => {
    getExamResult(testResult);
  };
  //#endregion

  return (
    <div>
      <Drawer anchor={"left"} open={openDrawer} onClose={toggleDrawer}>
        <ListAnswerResult
          userAnswers={userAnswers}
          loggedUser={loggedUser}
          onCreateNewNote={onCreateNewNote}
        />
      </Drawer>
      <Snackbar
        open={open}
        autoHideDuration={6000}
        onClose={handleClose}
        anchorOrigin={{ vertical: "top", horizontal: "center" }}
      >
        <Alert onClose={handleClose} severity="success" sx={{ width: "100%" }}>
          Submit successfully!
        </Alert>
      </Snackbar>
      {isError ? <div className="p-2 m-2 text-danger">{erroMsg}</div> : null}
      {isLoading ? (
        <p>
          <em>Loading...</em>
        </p>
      ) : (
        <div>
          <h3>Questions</h3>
          {data &&
            data.map((q: Questions) => (
              <div key={q.id} className="mb-2">
                <h6>{q.title}</h6>
                <p
                  dangerouslySetInnerHTML={{ __html: replaceWithBr(q) }}
                  style={{ fontStyle: "italic" }}
                ></p>
                <QuestionAnswerReadingDetails
                  key={q.id}
                  question={q}
                  questionDetails={q.questionDetails}
                  handleFieldChange={handleFieldChange}
                />
              </div>
            ))}

          <button
            className="btn btn-primary py-2 px-3 mr-2"
            type="submit"
            onClick={onSave}
            disabled={!isAuthenticated ? true : isSave ? true : false}
          >
            Submit
          </button>
          <button
            className="btn btn-primary py-2 px-3 mr-2"
            type="button"
            onClick={toggleDrawer}
            disabled={!isAuthenticated ? true : !isSave ? true : false}
          >
            View Results
          </button>
          <button
            className="btn btn-primary py-2 px-3 mr-2"
            type="button"
            onClick={restartNewTest}
            disabled={!isAuthenticated ? true : !isSave ? true : false}
          >
            Restart
          </button>
        </div>
      )}
    </div>
  );
};
export default QuestionAnswerReading;
