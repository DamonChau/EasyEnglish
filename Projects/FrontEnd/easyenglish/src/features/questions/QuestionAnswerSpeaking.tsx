/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
import React from "react";
import { useState, useEffect, useRef } from "react";
import { useTypedSelector } from "../../services";
import { selectLoggedUser } from "../../services/slices/authSlice";
import { UserAnswers } from "../../interfaces/interfaces";
import isUUID from "validator/lib/isUUID";
import Snackbar from "@mui/material/Snackbar";
import { useGetQuestionsWithQDQuery } from "./questionsApi";
import {
  useAddUserAnswerMutation,
  useUploadFilesMutation,
  useDownloadFilesMutation,
} from "../users/userAnswersApi";
import { useAddExamResultMutation } from "../examResult/examResultApi";
import { v4 as uuidv4 } from "uuid";
import Paper from "@mui/material/Paper";
import { styled } from "@mui/material/styles";
import { Alert } from "../common/Modals";
import {
  isFetchBaseQueryError,
  isErrorWithMessage,
} from "../../services/helpers";

const Item = styled(Paper)(({ theme }) => ({
  backgroundColor: "transparent",
  boxShadow: "none",
  ...theme.typography.body2,
  padding: theme.spacing(1),
  textAlign: "center",
  color: theme.palette.text.secondary,
}));

const QuestionAnswerSpeaking = ({ testId }: any) => {
  const [isView, setView] = React.useState(false);
  const { data, isFetching, isLoading, isSuccess, isError, error } =
    useGetQuestionsWithQDQuery(testId, { skip: !isView });
  const [
    addUserAnswer,
    { isLoading: isAddUALoading, isError: isAddUAError, error: errorUAAdd },
  ] = useAddUserAnswerMutation();
  const [
    uploadFile,
    { isLoading: isAddUFLoading, isError: isAddUFError, error: errorUFAdd },
  ] = useUploadFilesMutation();
  const [
    downloadFile,
    { isLoading: isAddDFLoading, isError: isAddDFError, error: errorDFAdd },
  ] = useDownloadFilesMutation();
  const [
    addExamResult,
    { isLoading: isAddERLoading, isError: isAddERError, error: errorERAdd },
  ] = useAddExamResultMutation();

  const loggedUser = useTypedSelector(selectLoggedUser);
  const initialValueTestResult = {
    id: uuidv4(),
    examTestId: testId,
    score: -1,
    noQuestion: 1,
    createdBy: loggedUser!.id,
  };
  const [testResult, setTestResult] = useState(initialValueTestResult);

  const initialValue = {
    id: uuidv4(),
    answer: "",
    result: -1,
    userId: loggedUser!.id,
    questionDetailId: uuidv4(),
    status: 1,
    examResultId: uuidv4(),
  };
  const [userAnswer, setUserAnswer] =
    useState<Partial<UserAnswers>>(initialValue);

  const [open, setOpen] = React.useState(false);

  const [permission, setPermission] = useState(false);
  const mediaRecorder = useRef<MediaRecorder | null>(null);
  const [recordingStatus, setRecordingStatus] = useState("inactive");
  const [stream, setStream] = useState<MediaStream>(new MediaStream());
  const [audio, setAudio] = useState("");
  const [audioChunks, setAudioChunks] = useState<Blob[]>([]);
  const [blobData, setBlobData] = useState<Blob>();
  const [erroMsg, setErrorMsg] = useState("");

  //update exam test result id
  useEffect(() => {
    if (testId && isUUID(testId)) {
      setView(true);
      setTestResult((prevState) => ({ ...prevState, examTestId: testId }));
    }
  }, [testId]);

  useEffect(() => {
    if (error) {
      if ("status" in error) {
        // you can access all properties of `FetchBaseQueryError` here
        const msg = "error" in error ? error.error : JSON.stringify(error.data);
        setErrorMsg(msg);
      } else {
        // you can access all properties of `SerializedError` here
        setErrorMsg(error.message as string);
      }
    }
  }, [isError]);

  const handleClose = (
    event?: React.SyntheticEvent | Event,
    reason?: string
  ) => {
    if (reason === "clickaway") {
      return;
    }
    setOpen(false);
  };

  const getMicrophonePermission = async () => {
    if ("MediaRecorder" in window) {
      try {
        const mediaStream = await navigator.mediaDevices.getUserMedia({
          audio: true,
          video: false,
        });
        setPermission(true);
        setStream(mediaStream as MediaStream);
      } catch (err: any) {
        alert(err.message);
      }
    } else {
      alert("The MediaRecorder API is not supported in your browser.");
    }
  };

  const startRecording = async () => {
    setRecordingStatus("recording");
    const options = {
      mimeType: "audio/webm",
    };
    const media: MediaRecorder = new MediaRecorder(stream, options);

    mediaRecorder.current! = media;

    mediaRecorder.current.start();

    const localAudioChunks: Blob[] = [];

    mediaRecorder.current.ondataavailable = (event) => {
      if (typeof event.data === "undefined") return;
      if (event.data.size === 0) return;
      localAudioChunks.push(event.data);
    };

    setAudioChunks(localAudioChunks);
  };

  const stopRecording = () => {
    setRecordingStatus("inactive");
    const options = {
      type: "audio/webm",
    };

    mediaRecorder.current!.stop();

    mediaRecorder.current!.onstop = () => {
      const audioBlob = new Blob(audioChunks, options);
      const audioUrl = URL.createObjectURL(audioBlob);
      setBlobData(audioBlob);
      setAudio(audioUrl);
      setAudioChunks([]);
    };
  };

  const handleSubmit = async () => {
    try {
      if (data) {
        data.forEach(async (question) => {
          question.questionDetails.forEach(async (qd) => {
            const ts = await addExamResult(testResult).unwrap();

            const newFile = new FormData();
            const fileName = ts.id + ".weba";
            newFile.append("file", blobData as Blob, fileName);
            await uploadFile(newFile).unwrap();

            const updatedUserAnswer = {
              ...userAnswer,
              questionDetailId: qd.id,
              examResultId: ts.id,
              answer: fileName,
            };
            setUserAnswer(updatedUserAnswer);
            await addUserAnswer(updatedUserAnswer).unwrap();
          });
        });
        setOpen(true);
      }
    } catch (err) {
      if (isFetchBaseQueryError(err)) {
        const msg =
          "error" in err
            ? err.error
            : JSON.parse(JSON.stringify(err.data)).error;
        setErrorMsg(msg);
      } else if (isErrorWithMessage(err)) console.log(err.message);
      setOpen(false);
    }
  };

  return (
    <React.Fragment>
      <Snackbar
        open={open}
        autoHideDuration={6000}
        onClose={handleClose}
        anchorOrigin={{ vertical: "top", horizontal: "center" }}
      >
        <Alert onClose={handleClose} severity="success" sx={{ width: "100%" }}>
          Submit successfully!
        </Alert>
      </Snackbar>
      {isLoading ? (
        <p>
          <em>Loading...</em>
        </p>
      ) : (
        <div className="card rounded-12 w-50">
          <h5 className="card-header h5">Recoring Your Answer</h5>
          <div className="card-body">
            <div className="row d-flex justify-content-left mb-2">
              {erroMsg ? (
                <div className="p-2 m-2 text-danger">{erroMsg}</div>
              ) : null}
              <div className="col-md-8 col-lg-12">
                {audio ? (
                  <audio
                    src={audio}
                    controls
                    className="question-answer-speaking-audio"
                  ></audio>
                ) : null}
              </div>
            </div>
            <div className="row d-flex justify-content-left">
              <div className="col-md-8 col-lg-12">
                {!permission ? (
                  <button
                    className="btn btn-primary py-1 px-2"
                    onClick={getMicrophonePermission}
                    type="button"
                  >
                    Get Microphone
                  </button>
                ) : null}

                {permission && recordingStatus === "inactive" ? (
                  <button
                    className="btn btn-primary py-1 px-2"
                    onClick={startRecording}
                    type="button"
                  >
                    Start Recording
                  </button>
                ) : null}
                {recordingStatus === "recording" ? (
                  <button
                    className="btn btn-primary py-1 px-2"
                    onClick={stopRecording}
                    type="button"
                  >
                    Stop Recording
                  </button>
                ) : null}
                {audio ? (
                  <button
                    className="btn btn-primary py-1 px-2 ml-2"
                    type="button"
                    onClick={handleSubmit}
                  >
                    Save
                  </button>
                ) : null}
              </div>
            </div>
          </div>
        </div>
      )}
    </React.Fragment>
  );
};

export default QuestionAnswerSpeaking;
