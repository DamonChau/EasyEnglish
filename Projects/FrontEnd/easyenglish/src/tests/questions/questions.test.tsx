/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-debugger */
/* eslint-disable @typescript-eslint/no-unused-vars */
//mock media stream object
Object.defineProperty(window, "MediaStream", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      start: jest.fn(),
      ondataavailable: jest.fn(),
      onerror: jest.fn(),
      state: "",
      stop: jest.fn(),
      pause: jest.fn(),
      resume: jest.fn(),
    })),
  });

  Object.defineProperty(MediaStream, "isTypeSupported", {
    writable: true,
    value: () => true,
  });
import "whatwg-fetch";
import React from "react";
import { waitFor, screen, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { renderWithProviders, store } from "../test-utils";
import QuestionManager from "../../features/questions/QuestionManager";
import QuestionAnswerReading from "../../features/questions/QuestionAnswerReading";
import QuestionAnswerSpeaking from "../../features/questions/QuestionAnswerSpeaking";
import QuestionAnswerWriting from "../../features/questions/QuestionAnswerWriting";
import { createStore } from "../../services/index";
import { mockQuestions } from "../questions/mock";
import { generateMockUser } from "../users/mock";

describe("Questions form", () => {
  test("QuestionManager Snapshot", () => {
    const item = renderWithProviders(<QuestionManager />, {
      store: createStore({
        preloadedState: {
          auth: {
            token: null,
            user: generateMockUser(),
            isAuthenticated: false,
            refreshToken: null,
          },
        },
      }),
    });
    expect(item).toMatchSnapshot();
  });
  test("QuestionAnswerReading Snapshot", () => {
    const onGetExamResult = () => {};
    const item = renderWithProviders(
      <QuestionAnswerReading
        testId="20312979-0936-4EAE-BE67-568C6652327E"
        onGetExamResult={onGetExamResult}
      />,
      {
        store: createStore({
          preloadedState: {
            auth: {
              token: null,
              user: generateMockUser(),
              isAuthenticated: false,
              refreshToken: null,
            },
          },
        }),
      }
    );
    expect(item).toMatchSnapshot();
  });
  test("QuestionAnswerSpeaking Snapshot", () => {
    const item = renderWithProviders(
      <QuestionAnswerSpeaking testId="20312979-0936-4EAE-BE67-568C6652327E" />,
      {
        store: createStore({
          preloadedState: {
            auth: {
              token: null,
              user: generateMockUser(),
              isAuthenticated: false,
              refreshToken: null,
            },
          },
        }),
      }
    );
    expect(item).toMatchSnapshot();
  });
  test("QuestionAnswerWriting Snapshot", () => {
    const item = renderWithProviders(
      <QuestionAnswerWriting testId="20312979-0936-4EAE-BE67-568C6652327E" />,
      {
        store: createStore({
          preloadedState: {
            auth: {
              token: null,
              user: generateMockUser(),
              isAuthenticated: false,
              refreshToken: null,
            },
          },
        }),
      }
    );
    expect(item).toMatchSnapshot();
  });
});
