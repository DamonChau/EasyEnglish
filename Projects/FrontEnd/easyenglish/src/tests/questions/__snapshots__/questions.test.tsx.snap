// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions form QuestionAnswerReading Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          <em>
            Loading...
          </em>
        </p>
      </div>
    </div>
    <script
      defer=""
      nonce="undefined"
      src="https://accounts.google.com/gsi/client"
    />
  </body>,
  "container": <div>
    <div>
      <p>
        <em>
          Loading...
        </em>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Questions form QuestionAnswerSpeaking Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p>
        <em>
          Loading...
        </em>
      </p>
    </div>
    <script
      defer=""
      nonce="undefined"
      src="https://accounts.google.com/gsi/client"
    />
  </body>,
  "container": <div>
    <p>
      <em>
        Loading...
      </em>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Questions form QuestionAnswerWriting Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p>
        <em>
          Loading...
        </em>
      </p>
    </div>
    <script
      defer=""
      nonce="undefined"
      src="https://accounts.google.com/gsi/client"
    />
  </body>,
  "container": <div>
    <p>
      <em>
        Loading...
      </em>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`Questions form QuestionManager Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <section
          class="hero-wrap hero-wrap-2"
          data-stellar-background-ratio="0.5"
          style="background-image: url(images/bg_1.jpg);"
        >
          <div
            class="overlay"
          />
          <div
            class="container"
          >
            <div
              class="row no-gutters slider-text align-items-center justify-content-center"
            >
              <div
                class="col-md-9 text-center"
              >
                <h1
                  class="mb-2 bread"
                >
                  Question Manager
                </h1>
                <p
                  class="breadcrumbs"
                >
                  <span
                    class="mr-2"
                  >
                    <a
                      href="/"
                    >
                      Home 
                      <i
                        class="ion-ios-arrow-forward"
                      />
                    </a>
                  </span>
                   
                  <span>
                    Questions Manager
                    <i
                      class="ion-ios-arrow-forward"
                    />
                  </span>
                </p>
              </div>
            </div>
          </div>
        </section>
        <section
          class="ftco-section"
        >
          <div
            class="container"
          >
            <div
              class="row"
            >
              <div
                class="col-lg-10"
              >
                <h2
                  class="mb-3"
                >
                  Question Manager
                </h2>
                <div>
                  <p>
                    <em>
                      Loading...
                    </em>
                  </p>
                </div>
              </div>
              <div
                class="col-lg-2 sidebar ftco-animate"
              />
            </div>
          </div>
        </section>
      </div>
    </div>
    <script
      defer=""
      nonce="undefined"
      src="https://accounts.google.com/gsi/client"
    />
  </body>,
  "container": <div>
    <div>
      <section
        class="hero-wrap hero-wrap-2"
        data-stellar-background-ratio="0.5"
        style="background-image: url(images/bg_1.jpg);"
      >
        <div
          class="overlay"
        />
        <div
          class="container"
        >
          <div
            class="row no-gutters slider-text align-items-center justify-content-center"
          >
            <div
              class="col-md-9 text-center"
            >
              <h1
                class="mb-2 bread"
              >
                Question Manager
              </h1>
              <p
                class="breadcrumbs"
              >
                <span
                  class="mr-2"
                >
                  <a
                    href="/"
                  >
                    Home 
                    <i
                      class="ion-ios-arrow-forward"
                    />
                  </a>
                </span>
                 
                <span>
                  Questions Manager
                  <i
                    class="ion-ios-arrow-forward"
                  />
                </span>
              </p>
            </div>
          </div>
        </div>
      </section>
      <section
        class="ftco-section"
      >
        <div
          class="container"
        >
          <div
            class="row"
          >
            <div
              class="col-lg-10"
            >
              <h2
                class="mb-3"
              >
                Question Manager
              </h2>
              <div>
                <p>
                  <em>
                    Loading...
                  </em>
                </p>
              </div>
            </div>
            <div
              class="col-lg-2 sidebar ftco-animate"
            />
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
